clippy_lints/src/lib.rs

```diff
@@ -348,6 +348,7 @@ pub fn register_pre_expansion_lints(store: &mut rustc_lint::LintStore) {
     store.register_pre_expansion_pass(|| box write::Write::default());
     store.register_pre_expansion_pass(|| box attrs::EarlyAttributes);
     store.register_pre_expansion_pass(|| box dbg_macro::DbgMacro);
+    store.register_pre_expansion_pass(|| box eq_op::EqOp);
```

Comment:
I've not done an in-depth review, but we are avoiding pre-expansion lints because they caused problems lately and support may be removed upstream AFAIK.

I think we should target the expanded version of the macros, so allow code from expansion and then somewhat check if it's from the list of macros we want to lint. I'm on mobile now so can't suggest how to do that, I'll take a look later.

Reply:
> support may be removed upstream AFAIK
Oh, didn't know about that.

I'm more worried about args comparison. The problem with expanded code seems that you can't check you are passing the same arguments.
```
assert_eq!(a, a);
```
is expanded into something like:
```
if left == right {
    ...
}
```
Hmm.. Maybe from `LateLintPass` we can get some ids from pre-expansion stage and compare them, I don't know if it will work though.

Reply:
You are right, this would need a bit more of work but I think it's doable. For example this code:
```rust
fn main() {
    let a = 42;
    assert_eq!(a, a);
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c85cfbc0f665621ac275b7c1265f6297))

expands to:
```rust
// ...
fn main() {
    let a = 42;
    {
        match (&a, &a) {
            (left_val, right_val) => {
                if !(*left_val == *right_val) {
                    // panicking code
                }
            }
        }
    };
}
```
We should be able to see that `left_val` and `right_val` are both utimately refering to `a`.

To check if the code comes from one of the macros we want to lint, we can use `utils::is_expn_of`.

Reply:
About pre-expansion lints, I learnt about it in #5518. There is an informal comment speaking about it [here](https://github.com/rust-lang/rust/pull/69838#issuecomment-599401121).

In general, I think using pre-expansion code does not work well because we would miss stuff like:
```rust
macro_rules! my_macro {
    () => {
        let a = 42;
        assert_eq!(a, a);
    }
}

fn main() {
    my_macro!();
}
```